File: tests/Unit/Evolution/Systems/CurvedScalarWave/BoundaryConditions/ConstraintPreservingSphericalRadiation.py
# Distributed under the MIT License.
# See LICENSE.txt for details.

import numpy as np
from Evolution.Systems.CurvedScalarWave.Characteristics import (
    char_speed_vminus,
    char_speed_vplus,
    char_speed_vpsi,
    char_speed_vzero,
)


def error(
    face_mesh_velocity,
    normal_covector,
    normal_vector,
    psi,
    phi,
    inertial_coords,
    gamma1,
    gamma2,
    lapse,
    shift,
    dt_psi,
    dt_pi,
    dt_phi,
    d_psi,
    d_pi,
    d_phi,
):
    return None


def subtract_mesh_velocity(
    face_mesh_velocity, dt_psi, dt_pi, dt_phi, d_psi, d_pi, d_phi
):
    if not (face_mesh_velocity is None):
        dt_psi -= np.einsum("i,i->", face_mesh_velocity, d_psi)
        dt_pi -= np.einsum("i,i->", face_mesh_velocity, d_pi)
        dt_phi -= np.einsum("i,ij->j", face_mesh_velocity, d_phi)


def dt_psi_constraint_preserving_spherical_radiation(
    face_mesh_velocity,
    normal_covector,
    normal_vector,
    psi,
    phi,
    inertial_coords,
    gamma1,
    gamma2,
    lapse,
    shift,
    dt_psi,
    dt_pi,
    dt_phi,
    d_psi,
    d_pi,
    d_phi,
):
    subtract_mesh_velocity(
        face_mesh_velocity, dt_psi, dt_pi, dt_phi, d_psi, d_pi, d_phi
    )
    char_speed_psi = char_speed_vpsi(gamma1, lapse, shift, normal_covector)

    if face_mesh_velocity is not None:
        char_speed_psi -= np.dot(normal_covector, face_mesh_velocity)

    return np.dot(normal_vector, d_psi - phi) * min(0.0, char_speed_psi)


def dt_phi_constraint_preserving_spherical_radiation(
    face_mesh_velocity,
    normal_covector,
    normal_vector,
    psi,
    phi,
    inertial_coords,
    gamma1,
    gamma2,
    lapse,
    shift,
    dt_psi,
    dt_pi,
    dt_phi,
    d_psi,
    d_pi,
    d_phi,
):
    subtract_mesh_velocity(
        face_mesh_velocity, dt_psi, dt_pi, dt_phi, d_psi, d_pi, d_phi
    )
    char_speed_zero = char_speed_vzero(gamma1, lapse, shift, normal_covector)
    if face_mesh_velocity is not None:
        char_speed_zero -= np.dot(normal_covector, face_mesh_velocity)
    return np.einsum("ij,j", d_phi.T - d_phi, normal_vector) * min(
        0, char_speed_zero
    )


def dt_pi_constraint_preserving_spherical_radiation(
    face_mesh_velocity,
    normal_covector,
    normal_vector,
    psi,
    phi,
    inertial_coords,
    gamma1,
    gamma2,
    lapse,
    shift,
    dt_psi,
    dt_pi,
    dt_phi,
    d_psi,
    d_pi,
    d_phi,
):
    dt_psi_correction = dt_psi_constraint_preserving_spherical_radiation(
        face_mesh_velocity,
        normal_covector,
        normal_vector,
        psi,
        phi,
        inertial_coords,
        gamma1,
        gamma2,
        lapse,
        shift,
        dt_psi,
        dt_pi,
        dt_phi,
        d_psi,
        d_pi,
        d_phi,
    )
    # dt's are already corrected in dt_psi
    inv_radius = 1.0 / np.linalg.norm(inertial_coords)
    bc_dt_pi = (
        2.0 * inv_radius**2 * psi
        + 4.0 * inv_radius * dt_psi
        + 4.0 * inv_radius * np.dot(normal_vector, phi)
        + 2.0 * np.dot(normal_vector, dt_phi)
        + np.dot(shift, dt_phi)
        + np.einsum("i,j,ij", normal_vector, normal_vector, d_phi)
    )
    bc_dt_pi /= lapse
    return bc_dt_pi - dt_pi + gamma2 * dt_psi_correction

File: src/Evolution/Systems/CurvedScalarWave/BoundaryConditions/ConstraintPreservingSphericalRadiation.cpp
// Distributed under the MIT License.
// See LICENSE.txt for details.

#include "Evolution/Systems/CurvedScalarWave/BoundaryConditions/ConstraintPreservingSphericalRadiation.hpp"

#include <cstddef>
#include <memory>
#include <optional>
#include <pup.h>

#include "DataStructures/DataVector.hpp"
#include "DataStructures/Tags/TempTensor.hpp"
#include "DataStructures/Tensor/EagerMath/DotProduct.hpp"
#include "DataStructures/Tensor/Tensor.hpp"
#include "DataStructures/Variables.hpp"
#include "Evolution/Systems/CurvedScalarWave/Characteristics.hpp"
#include "Options/ParseOptions.hpp"
#include "Utilities/GenerateInstantiations.hpp"

namespace CurvedScalarWave::BoundaryConditions {

template <size_t Dim>
std::unique_ptr<domain::BoundaryConditions::BoundaryCondition>
ConstraintPreservingSphericalRadiation<Dim>::get_clone() const {
  return std::make_unique<ConstraintPreservingSphericalRadiation>(*this);
}

template <size_t Dim>
void ConstraintPreservingSphericalRadiation<Dim>::pup(PUP::er& p) {
  BoundaryCondition<Dim>::pup(p);
}

template <size_t Dim>
ConstraintPreservingSphericalRadiation<
    Dim>::ConstraintPreservingSphericalRadiation(CkMigrateMessage* const msg)
    : BoundaryCondition<Dim>(msg) {}

template <size_t Dim>
std::optional<std::string>
ConstraintPreservingSphericalRadiation<Dim>::dg_time_derivative(
    const gsl::not_null<Scalar<DataVector>*> dt_pi_correction,
    const gsl::not_null<tnsr::i<DataVector, Dim, Frame::Inertial>*>
        dt_phi_correction,
    const gsl::not_null<Scalar<DataVector>*> dt_psi_correction,
    const std::optional<tnsr::I<DataVector, Dim, Frame::Inertial>>&
        face_mesh_velocity,
    const tnsr::i<DataVector, Dim>& normal_covector,
    const tnsr::I<DataVector, Dim>& normal_vector,
    const tnsr::i<DataVector, Dim>& phi, const Scalar<DataVector>& psi,
    const tnsr::I<DataVector, Dim, Frame::Inertial>& coords,
    const Scalar<DataVector>& gamma1, const Scalar<DataVector>& gamma2,
    const Scalar<DataVector>& lapse, const tnsr::I<DataVector, Dim>& shift,
    const Scalar<DataVector>& dt_pi, const tnsr::i<DataVector, Dim>& dt_phi,
    const Scalar<DataVector>& dt_psi, const tnsr::i<DataVector, Dim>& d_psi,
    const tnsr::ij<DataVector, Dim>& d_phi) const {
  Variables<tmpl::list<::Tags::Tempa<0, 3>, ::Tags::TempScalar<1>,
                       ::Tags::TempScalar<2>>>
      buffer{get<0>(coords).size()};

  auto& char_speeds = get<::Tags::Tempa<0, 3>>(buffer);
  auto& inv_radius = get<::Tags::TempScalar<1>>(buffer);
  auto& face_speed = get<::Tags::TempScalar<2>>(buffer);

  get(inv_radius) = 1. / get(magnitude(coords));
  characteristic_speeds(make_not_null(&char_speeds), gamma1, lapse, shift,
                        normal_covector);

  if (face_mesh_velocity.has_value()) {
    face_speed = dot_product(normal_covector, *face_mesh_velocity);
    for (size_t i = 0; i < 4; ++i) {
      char_speeds.get(i) -= get(face_speed);
    }
  }

  get(*dt_psi_correction) =
      get<0>(normal_vector) * (get<0>(d_psi) - get<0>(phi));
  for (size_t i = 1; i < Dim; ++i) {
    get(*dt_psi_correction) +=
        normal_vector.get(i) * (d_psi.get(i) - phi.get(i));
  }

  for (size_t i = 0; i < Dim; ++i) {
    dt_phi_correction->get(i) =
        get<0>(normal_vector) * (d_phi.get(0, i) - d_phi.get(i, 0));
    for (size_t j = 1; j < Dim; ++j) {
      dt_phi_correction->get(i) +=
          normal_vector.get(j) * (d_phi.get(j, i) - d_phi.get(i, j));
    }
  }

  for (size_t i = 0; i < get(*dt_psi_correction).size(); ++i) {
    get(*dt_psi_correction)[i] *= std::min(0., gsl::at(char_speeds[0], i));
    for (size_t j = 0; j < Dim; ++j) {
      dt_phi_correction->get(j)[i] *=
          0.5 * std::min(0., gsl::at(char_speeds[1], i));
    }
  }

  get(*dt_pi_correction) =
      (2.0 * get(inv_radius) * get(psi) + 4.0 * get(dt_psi)) * get(inv_radius);
  for (size_t i = 0; i < Dim; ++i) {
    get(*dt_pi_correction) +=
        normal_vector.get(i) *
            (2. * dt_phi.get(i) + 4.0 * get(inv_radius) * phi.get(i)) +
        shift.get(i) * dt_phi.get(i);
    for (size_t j = 0; j < Dim; ++j) {
      get(*dt_pi_correction) +=
          normal_vector.get(i) * normal_vector.get(j) * d_phi.get(i, j);
    }
  }
  get(*dt_pi_correction) /= get(lapse);
  get(*dt_pi_correction) += get(gamma2) * get(*dt_psi_correction) - get(dt_pi);
  return {};
}

template <size_t Dim>
// NOLINTNEXTLINE
PUP::able::PUP_ID ConstraintPreservingSphericalRadiation<Dim>::my_PUP_ID = 0;

#define DIM(data) BOOST_PP_TUPLE_ELEM(0, data)

#define INSTANTIATION(r, data) \
  template class ConstraintPreservingSphericalRadiation<DIM(data)>;

GENERATE_INSTANTIATIONS(INSTANTIATION, (1, 2, 3))

#undef INSTANTIATION
#undef DIM
}  // namespace CurvedScalarWave::BoundaryConditions
